### PR TITLE
[BUGFIX] Reset Metrics Registry in order to keep state of Metrics Registry test cases Runs mutually consistent

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Scope
* Reset Metrics Registry at each test case run (otherwise, subsequent tests keep updating it, making assertions inconsistent).
* Clean up unused type hint.

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

Changes proposed in this pull request:
- JIRA: GREAT-761/GREAT-1285/DX-392
-
-

After submitting your PR, CI checks will run and @cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: tests/expectations/metrics/test_metric_providers.py [hunk 1]
```diff
@@ -29,17 +30,22 @@
...
-)
+from great_expectations.expectations import registry
+
+
+_REGISTERED_METRICS_ORIGINAL: dict = copy.deepcopy(registry._registered_metrics)
 
 
 def test__base_metric_provider__registration():
     """This tests whether the MetricProvider class registers the correct metrics."""
+    registry._registered_metrics = copy.deepcopy(
```
Comment (billdirks) [inline tests/expectations/metrics/test_metric_providers.py]: Could you use [monkeypatch setattr](https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest.MonkeyPatch.setattr)?

My concern here is this doesn't reset this attribute after each test. That means, tests that don't test the registry but depend on it indirectly can fail if one of these tests fails part way through. So, I'd like to see monkeypatch or a fixture that unrolls the change.

**Review by billdirks — CHANGES_REQUESTED** [review]
Thanks for identifying and fixing the issue. I left an implementation comment inline.

File: tests/expectations/metrics/test_metric_providers.py (hunk 1 above)
Comment (Kilo59) [reply tests/expectations/metrics/test_metric_providers.py]: You can factor this into a fixture with a setup and teardown step. You can save whatever state you need to at the beginning of the fixture and reset it during the teardown. Or just let monkeypatch do that part for you.

https://docs.pytest.org/en/7.1.x/how-to/fixtures.html#teardown-cleanup-aka-fixture-finalization

Comment (alexsherstinsky) [reply tests/expectations/metrics/test_metric_providers.py]: @billdirks Done -- please review.  Thanks.

File: tests/expectations/metrics/test_metric_providers.py [hunk 2]
```diff
@@ -29,17 +34,26 @@
...
 )
-from great_expectations.expectations.registry import (
-    _registered_metrics,
-)
+from great_expectations.expectations import registry
 
 
-def test__base_metric_provider__registration():
-    """This tests whether the MetricProvider class registers the correct metrics."""
+_REGISTERED_METRICS_ORIGINAL: dict = copy.deepcopy(registry._registered_metrics)
```
Comment (Kilo59) [inline tests/expectations/metrics/test_metric_providers.py]: You shouldn't need this global.
You can do this in the setup section of a fixture.

Comment (alexsherstinsky) [reply tests/expectations/metrics/test_metric_providers.py]: @Kilo59 Please take a look at the latest -- thank you.

File: tests/expectations/metrics/test_metric_providers.py [hunk 3]
```diff
@@ -29,17 +34,27 @@
...
-    _registered_metrics,
-)
+from great_expectations.expectations import registry
 
 
-def test__base_metric_provider__registration():
-    """This tests whether the MetricProvider class registers the correct metrics."""
+@pytest.fixture
+def mock_registry(monkeypatch: pytest.MonkeyPatch):
+    # ensures consistent state among test cases runs
```
Comment (Kilo59) [inline tests/expectations/metrics/test_metric_providers.py]:
```suggestion
    """Ensures consistent `_registered_metrics` state among test cases runs."""
```
Might as well make this a docstring

File: tests/expectations/metrics/test_metric_providers.py [hunk 4]
```diff
@@ -29,17 +34,27 @@
...
+    monkeypatch.setattr(
+        registry,
+        "_registered_metrics",
+        copy.deepcopy(registry._registered_metrics),
+        raising=False,
```
Comment (Kilo59) [inline tests/expectations/metrics/test_metric_providers.py]:
```suggestion
        raising=True,
```
This ensures monkeypatch raises an error if this attribute doesn't exist.
Which we want. We don't want this to keep working if we change where the `_registered_metrics` are stored.
https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest.MonkeyPatch.setattr

**Comment by alexsherstinsky** [comment]
Thank you, @Kilo59! (both suggestions committed -- straight from within GitHub).

**Review by Kilo59 — APPROVED** [review]

**Review by billdirks — APPROVED** [review]
LGTM!

**Pull request merged**